File: go/tcp_test.go
// Copyright (c) SandboxAQ. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-only

package sandwich_test

import (
	"github.com/sandbox-quantum/sandwich/go"
	"crypto/rand"
	"fmt"
	"math/big"
	"sync"
	"testing"

	"github.com/bazelbuild/rules_go/go/tools/bazel"

	api "github.com/sandbox-quantum/sandwich/go/proto/sandwich/api/v1"
)

var testCertPattern = "testdata/%s.cert.pem"

var (
	pingMsg         = [...]byte{'P', 'I', 'N', 'G'}
	pongMsg         = [...]byte{'P', 'O', 'N', 'G'}
	certPath string = "testdata/localhost.cert.pem"
	keyPath  string = "testdata/localhost.key.pem"
)

// createServerConfiguration creates the configuration for the server.
func createServerConfiguration(t *testing.T) (*api.Configuration, error) {
	certfile, err := bazel.Runfile(certPath)
	if err != nil {
		t.Errorf("Could not load certificate file %s: %v", certPath, err)
	}
	keyfile, err := bazel.Runfile(keyPath)
	if err != nil {
		t.Errorf("Could not load private key file %s: %v", keyPath, err)
	}

	return &api.Configuration{
		Impl: api.Implementation_IMPL_OPENSSL1_1_1_OQS,
		Opts: &api.Configuration_Server{
			Server: &api.ServerOptions{
				Opts: &api.ServerOptions_Tls{
					Tls: &api.TLSServerOptions{
						CommonOptions: &api.TLSOptions{
							Tls13: &api.TLSv13Config{
								Ke: []string{
									"kyber1024",
								},
							},
							PeerVerifier: &api.TLSOptions_EmptyVerifier{
								EmptyVerifier: &api.EmptyVerifier{},
							},
							Identity: &api.X509Identity{
								Certificate: &api.Certificate{
									Source: &api.Certificate_Static{
										Static: &api.ASN1DataSource{
											Data: &api.DataSource{
												Specifier: &api.DataSource_Filename{
													Filename: certfile,
												},
											},
											Format: api.ASN1EncodingFormat_ENCODING_FORMAT_PEM,
										},
									},
								},
								PrivateKey: &api.PrivateKey{
									Source: &api.PrivateKey_Static{
										Static: &api.ASN1DataSource{
											Data: &api.DataSource{
												Specifier: &api.DataSource_Filename{
													Filename: keyfile,
												},
											},
											Format: api.ASN1EncodingFormat_ENCODING_FORMAT_PEM,
										},
									},
								},
							},
						},
					},
				},
			},
		},
	}, nil
}

// createClientConfiguration creates the configuration for the client.
func createClientConfiguration(t *testing.T) (*api.Configuration, error) {
	certfile, err := bazel.Runfile(certPath)
	if err != nil {
		t.Errorf("Could not load certificate file %s: %v", certPath, err)
	}

	return &api.Configuration{
		Impl: api.Implementation_IMPL_OPENSSL1_1_1_OQS,
		Opts: &api.Configuration_Client{
			Client: &api.ClientOptions{
				Opts: &api.ClientOptions_Tls{
					Tls: &api.TLSClientOptions{
						CommonOptions: &api.TLSOptions{
							Tls13: &api.TLSv13Config{
								Ke: []string{
									"kyber1024",
								},
							},
							PeerVerifier: &api.TLSOptions_X509Verifier{
								X509Verifier: &api.X509Verifier{
									TrustedCas: []*api.Certificate{
										{
											Source: &api.Certificate_Static{
												Static: &api.ASN1DataSource{
													Data: &api.DataSource{
														Specifier: &api.DataSource_Filename{
															Filename: certfile,
														},
													},
													Format: api.ASN1EncodingFormat_ENCODING_FORMAT_PEM,
												},
											},
										},
									},
								},
							},
						},
					},
				},
			},
		},
	}, nil
}

// createServerContext creates the server context.
func createServerContext(t *testing.T, sw *sandwich.Sandwich) (*sandwich.TunnelContext, error) {
	config, err := createServerConfiguration(t)
	if err != nil {
		t.Errorf("Failed to create the server configuration: %v", err)
		panic("failed")
	}

	ctx, err := sandwich.NewTunnelContext(sw, config)
	if err != nil {
		t.Errorf("Failed to create the server context: %v", err)
		panic("failed")
	}

	return ctx, nil
}

// createClientContext creates the client context.
func createClientContext(t *testing.T, sw *sandwich.Sandwich) (*sandwich.TunnelContext, error) {
	config, err := createClientConfiguration(t)
	if err != nil {
		t.Errorf("Failed to create the client configuration: %v", err)
		panic("failed")
	}

	ctx, err := sandwich.NewTunnelContext(sw, config)
	if err != nil {
		t.Errorf("Failed to create the client context: %v", err)
	}

	return ctx, nil
}

type ioInts struct {
	client sandwich.TunnelIO
	server sandwich.TunnelIO
}

func generateRandomPort() uint16 {
	randNum, err := rand.Int(rand.Reader, big.NewInt(64510))
	if err != nil {
		return 0
	}
	return uint16(randNum.Int64() + 1026)
}

func createListenerConfiguration(ipaddr string, port uint16) *api.ListenerConfiguration {
	return &api.ListenerConfiguration{
		Mode: &api.ListenerConfiguration_Tcp{
			Tcp: &api.ListenerModeTCP{
				Addr: &api.SocketAddress{
					Hostname: ipaddr,
					Port:     uint32(port),
				},
				BlockingMode: api.BlockingMode_BLOCKINGMODE_BLOCKING,
			},
		},
	}
}

// createServerClientIOs creates the I/O interfaces for the server and the client.
func createIOs() ioInts {
	hostname := "127.0.0.1"
	port := generateRandomPort()
	listener_config := createListenerConfiguration(hostname, port)
	listener, err := sandwich.NewListener(listener_config)
	if err != nil {
		fmt.Println("Error listening:", err)
	}
	listener.Listen()
	client, _ := sandwich.IOTCPClient(hostname, port, true)
	server, err2 := listener.Accept()
	if err2 != nil {
		fmt.Println("Error accepting:", err)
	}
	return ioInts{
		client: client,
		server: server,
	}
}

// createServerTunnel creates the tunnel for the server.
func createServerTunnel(t *testing.T, context *sandwich.TunnelContext, io sandwich.TunnelIO) (*sandwich.Tunnel, error) {
	tun, err := sandwich.NewTunnel(context, io, createTunnelConfigurationWithEmptyTunnelVerifier())
	if err != nil {
		t.Errorf("Failed to create the server's tunnel: %v", err)
	}

	return tun, nil
}

// createClientTunnel creates the tunnel for the client.
func createClientTunnel(t *testing.T, context *sandwich.TunnelContext, io sandwich.TunnelIO) (*sandwich.Tunnel, error) {
	tun, err := sandwich.NewTunnel(context, io, createTunnelConfigurationWithEmptyTunnelVerifier())
	if err != nil {
		t.Errorf("Failed to create the client's tunnel: %v", err)
	}

	return tun, nil
}

// createTunnelConfigurationWithEmptyTunnelVerifier creates a tunnel configuration with an empty TunnelVerifier.
func createTunnelConfigurationWithEmptyTunnelVerifier() *api.TunnelConfiguration {
	return &api.TunnelConfiguration{
		Verifier: &api.TunnelVerifier{
			Verifier: &api.TunnelVerifier_EmptyVerifier{
				EmptyVerifier: &api.EmptyVerifier{},
			},
		},
	}
}

func clientRoutine(t *testing.T, wg *sync.WaitGroup, clientTunnel *sandwich.Tunnel, recvMsg [4]byte, sendMsg [4]byte) {
	var buf [4]byte
	err := clientTunnel.Handshake()
	n := 0
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}

	n, err = clientTunnel.Write(pingMsg[:])
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}
	if n != len(pingMsg) {
		t.Errorf("Expected %v bytes sent, got %v", len(pingMsg), n)
	}

	n, err = clientTunnel.Read(buf[:])
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}
	if n != len(pongMsg) {
		t.Errorf("Expected %v bytes read, got %v", len(pongMsg), n)
	}
	if buf != pongMsg {
		t.Errorf("Expected %v, got %v", pongMsg, buf)
	}

	clientTunnel.Close()
	wg.Done()
}

func serverRoutine(t *testing.T, wg *sync.WaitGroup, serverTunnel *sandwich.Tunnel, recvMsg [4]byte, sendMsg [4]byte) {
	var buf [4]byte
	n := 0
	err := serverTunnel.Handshake()
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}

	n, err = serverTunnel.Read(buf[:])
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}
	if n != len(recvMsg) {
		t.Errorf("Expected %v bytes read, got %v", len(recvMsg), n)
	}
	if buf != recvMsg {
		t.Errorf("Expected %v, got %v", recvMsg, buf)
	}

	n, err = serverTunnel.Write(sendMsg[:])
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}
	if n != len(sendMsg) {
		t.Errorf("Expected %v bytes sent, got %v", len(sendMsg), n)
	}

	serverTunnel.Close()
	wg.Done()
}

func TestTunnels(t *testing.T) {
	var wg sync.WaitGroup
	wg.Add(2)
	sw := sandwich.NewSandwich()
	serverContext, err := createServerContext(t, sw)
	if err != nil {
		t.Errorf("Failed to create Server context: %v", err)
	}

	clientContext, err := createClientContext(t, sw)
	if err != nil {
		t.Errorf("Failed to create Client context: %v", err)
	}

	ioInterfaces := createIOs()

	serverTunnel, err := createServerTunnel(t, serverContext, ioInterfaces.server)
	if err != nil {
		t.Errorf("Failed to create the server tunnel: %v", err)
	}

	clientTunnel, err := createClientTunnel(t, clientContext, ioInterfaces.client)
	if err != nil {
		t.Errorf("Failed to create the server tunnel: %v", err)
	}
	go func() {
		serverRoutine(t, &wg, serverTunnel, pingMsg, pongMsg)
	}()
	go func() {
		clientRoutine(t, &wg, clientTunnel, pongMsg, pingMsg)
	}()
	wg.Wait()
}

File: pysandwich/tcp_test.py
# Copyright (c) SandboxAQ. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-only
import random
import threading

import pysandwich.proto.api.v1.configuration_pb2 as SandwichAPI
import pysandwich.proto.api.v1.encoding_format_pb2 as EncodingFormat
import pysandwich.proto.api.v1.listener_configuration_pb2 as ListenerAPI
import pysandwich.proto.api.v1.verifiers_pb2 as SandwichVerifiers
import pysandwich.errors as errors
from pysandwich.proto.api.v1.tunnel_pb2 import TunnelConfiguration
from pysandwich import listener, tunnel
from pysandwich.io_helpers import io_client_tcp_new
from pysandwich.sandwich import Sandwich

_LISTENING_ADDRESS = "127.0.0.1"
_LISTENING_PORT = random.randint(1226, 65535)

_PING_MSG = b"PING"
_PONG_MSG = b"PONG"

_CERT_PATH = "testdata/falcon1024.cert.pem"
_KEY_PATH = "testdata/falcon1024.key.pem"

_DEFAULT_KE = "kyber512"


def create_server_conf(sw: Sandwich) -> tunnel.Context:
    """Creates the configuration for the server.

    Returns:
        Configuration for the server.
    """
    conf = SandwichAPI.Configuration()
    conf.impl = SandwichAPI.IMPL_OPENSSL1_1_1_OQS

    # Sets TLS 1.3 Compliance and Key Establishment (KE)
    tls13 = conf.server.tls.common_options.tls13
    tls13.ke.append(_DEFAULT_KE)

    conf.server.tls.common_options.empty_verifier.CopyFrom(
        SandwichVerifiers.EmptyVerifier()
    )
    conf.server.tls.common_options.identity.certificate.static.data.filename = (
        _CERT_PATH
    )
    conf.server.tls.common_options.identity.certificate.static.format = (
        EncodingFormat.ENCODING_FORMAT_PEM
    )

    conf.server.tls.common_options.identity.private_key.static.data.filename = _KEY_PATH
    conf.server.tls.common_options.identity.private_key.static.format = (
        EncodingFormat.ENCODING_FORMAT_PEM
    )

    return tunnel.Context.from_config(sw, conf)


def create_client_conf(sw: Sandwich) -> tunnel.Context:
    """Creates the configuration for the client.

    Returns:
        Configuration for the client.
    """
    conf = SandwichAPI.Configuration()
    conf.impl = SandwichAPI.IMPL_OPENSSL1_1_1_OQS

    # Sets TLS 1.3 Compliance and Key Establishment (KE)
    tls13 = conf.client.tls.common_options.tls13
    tls13.ke.append(_DEFAULT_KE)

    cert = conf.client.tls.common_options.x509_verifier.trusted_cas.add().static
    cert.data.filename = _CERT_PATH
    cert.format = EncodingFormat.ENCODING_FORMAT_PEM

    buf = conf.SerializeToString()
    return tunnel.Context.from_bytes(sw, buf)


def create_tcp_listener(hostname, port) -> listener.Listener:
    """Creates the configuration for a TCP listener.
    Returns:
        A TCP listener which is listening on hostname:port.
    """
    conf = ListenerAPI.ListenerConfiguration()
    conf.tcp.addr.hostname = hostname
    conf.tcp.addr.port = port
    conf.tcp.blocking_mode = ListenerAPI.BLOCKINGMODE_BLOCKING

    return listener.Listener(conf)


def server_thread(hostname, port, server_conf, exceptions, server_ready):
    if server_conf is None:
        # prevent the other thread from hanging
        server_ready.set()
        exceptions.append(AssertionError("server_conf is None"))
        return

    tunnel_configuration = TunnelConfiguration()
    tunnel_configuration.verifier.empty_verifier.CopyFrom(
        SandwichVerifiers.EmptyVerifier()
    )
    try:
        listener = create_tcp_listener(hostname, port)
        listener.listen()
    except Exception as e:
        # prevent the other thread from hanging
        server_ready.set()
        exceptions.append(e)
        return
    server_ready.set()
    server_io = None
    try:
        server_io = listener.accept()
    except Exception as e:
        exceptions.append(e)
        return
    server = tunnel.Tunnel(server_conf, server_io, tunnel_configuration)
    if server is None:
        exceptions.append(AssertionError("server is None"))
        return
    try:
        server.handshake()
    except Exception as e:
        exceptions.append(e)
        server.close()
        return
    state = server.state()
    if state != server.State.STATE_HANDSHAKE_DONE:
        exceptions.append(AssertionError(f"Expected state HANDSHAKE_DONE, got {state}"))
        server.close()
        return
    try:
        data = server.read(len(_PING_MSG))
    except errors.RecordPlaneException as e:
        exceptions.append(e)
        server.close()
        return
    if len(data) != len(_PING_MSG):
        exceptions.append(
            AssertionError(f"Expected {len(_PING_MSG)} bytes read, got {len(data)}")
        )
        server.close()
        return
    if data != _PING_MSG:
        exceptions.append(
            AssertionError(f"Expected msg {_PING_MSG} from client, got {data}")
        )
        server.close()
        return

    w = 0
    try:
        w = server.write(_PONG_MSG)
    except errors.RecordPlaneException as e:
        exceptions.append(e)
        server.close()
        return
    if w != len(_PONG_MSG):
        exceptions.append(
            AssertionError(f"Expected {len(_PING_MSG)} bytes written, got {w}")
        )
    server.close()


def client_thread(hostname, port, client_conf, exceptions, server_ready):
    if client_conf is None:
        exceptions.append(AssertionError("client_conf is None"))
        return

    tunnel_configuration = TunnelConfiguration()
    tunnel_configuration.verifier.empty_verifier.CopyFrom(
        SandwichVerifiers.EmptyVerifier()
    )
    server_ready.wait()
    client_io = None
    try:
        client_io = io_client_tcp_new(hostname, port, True)
    except Exception as e:
        exceptions.append(AssertionError(f"failed to create client_io: {e}"))
        return
    client = tunnel.Tunnel(client_conf, client_io, tunnel_configuration)
    if client is None:
        exceptions.append(AssertionError("client is None"))
        return

    try:
        client.handshake()
    except Exception as e:
        client.close()
        exceptions.append(e)
        return

    state = client.state()
    if state != client.State.STATE_HANDSHAKE_DONE:
        exceptions.append(AssertionError(f"Expected state HANDSHAKE_DONE, got {state}"))
        client.close()
        return

    w = 0
    try:
        w = client.write(_PING_MSG)
    except errors.RecordPlaneException as e:
        exceptions.append(e)
        client.close()
        return
    if w != len(_PING_MSG):
        exceptions.append(
            AssertionError(f"Expected {len(_PING_MSG)} bytes written, got {w}")
        )
        client.close()
        return

    try:
        data = client.read(len(_PONG_MSG))
    except errors.RecordPlaneException as e:
        exceptions.append(e)
        client.close()
        return
    if len(data) != len(_PONG_MSG):
        exceptions.append(
            AssertionError(f"Expected {len(_PING_MSG)} bytes read, got {len(data)}")
        )
        client.close()
        return
    if data != _PONG_MSG:
        exceptions.append(
            AssertionError(f"Expected msg {_PING_MSG} from client, got {data}")
        )
    client.close()


def main():
    sw = Sandwich()
    threads = []
    client_conf = create_client_conf(sw)
    assert client_conf is not None
    server_conf = create_server_conf(sw)
    assert server_conf is not None
    server_ready = threading.Event()
    client_exceptions = []
    client_args = (
        _LISTENING_ADDRESS,
        _LISTENING_PORT,
        client_conf,
        client_exceptions,
        server_ready,
    )
    server_exceptions = []
    server_args = (
        _LISTENING_ADDRESS,
        _LISTENING_PORT,
        server_conf,
        server_exceptions,
        server_ready,
    )
    threads.append(threading.Thread(target=client_thread, args=client_args))
    threads.append(threading.Thread(target=server_thread, args=server_args))

    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join(timeout=5)
    if client_exceptions != []:
        print("===========Client thread exceptions===========")
    for e in client_exceptions:
        print(e)
    if server_exceptions != []:
        print("===========Server thread exceptions===========")
    for e in server_exceptions:
        print(e)
    if client_exceptions != [] or server_exceptions != []:
        print("==============================================", flush=True)
        raise AssertionError("A child thread threw an exception")


if __name__ == "__main__":
    main()
